auth/realm/base/src/main/java/org/wildfly/security/auth/realm/FileSystemSecurityRealm.java

```diff
@@ -965,6 +966,8 @@ protected LoadedIdentity loadIdentityPrivileged(final boolean skipCredentials, f
                 } catch (XMLStreamException e) {
                     throw ElytronMessages.log.fileSystemRealmFailedToRead(path, name, e);
                 }
+            } catch (RealmUnavailableException e) {
+                throw ElytronMessages.log.fileSystemRealmInvalidSecretKey(e);
```

Comment:
We need to be careful here and below. With this approach, we'd always throw an exception indicating the secret key was invalid regardless of the actual cause of the `RealmUnvailableException`.

To make sure we only throw `fileSystemRealmInvalidSecretKey` when we actually have an invalid key, we should be wrapping the calls to `CipherUtil.encrypt` and `CipherUtil.decrypt` with a `try/catch` block. If we catch a `GeneralSecurityException` (which is the exception thrown by `encrypt` / `decrypt` if something has gone wrong), then we should throw `fileSystemRealmInvalidSecretKey`.

Reply:
Just to confirm my understanding. Right now it's throwing a `RealmUnvailableException` (using `fileSystemRealmInvalidSecretKey`) through all the `try/catch`'s from the lowest level to the highest, but instead we should be throwing a `GeneralSecurityException` all the way to the highest level where there the `GeneralSecurityException` should be caught and then rethrown as a `fileSystemRealmInvalidSecretKey`?

Reply:
The idea is to catch the `GeneralSecurityException` right after a call to `encrypt` / `decrypt` and to re-throw as a `RealmUnavailableException` with the `GeneralSecurityException` as the cause.